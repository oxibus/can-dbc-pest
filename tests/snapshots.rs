#![cfg(feature = "encoding")]

use std::borrow::Cow;
use std::fs;
use std::path::Path;

use can_dbc_pest::{decode_cp1252, DbcParser, Rule};
use insta::{assert_debug_snapshot, with_settings};
use pest::Parser;

/// Test parsing all DBC files
#[test]
fn test_cantools_dbc_files() {
    let test_dirs = [
        (
            "tests/shared-test-files/dbc-cantools",
            "dbc-cantools",
            true,
            true,
        ),
        ("tests/shared-test-files/canpy", "canpy", true, true),
        ("tests/opendbc/opendbc/dbc", "opendbc", false, false),
    ];
    for (path, file_name, snapshot, cp1251) in test_dirs {
        let path = Path::new(path);
        let snapshot_path = String::from("snapshots-") + file_name;
        // snapshots should go to /tests/snapshots-* directory
        with_settings! {
            { omit_expression => true,
              snapshot_path => snapshot_path ,
              prepend_module_to_snapshot => false },
            {
                test_dbc_files(path, snapshot, cp1251);
            }
        }
    }
}

/// Test parsing all DBC files in the given directory.
fn test_dbc_files(dir: impl AsRef<Path>, snapshot: bool, cp1251: bool) {
    let dir = dir.as_ref();
    let dir_display = dir.display();
    let dir_content = fs::read_dir(dir).unwrap_or_else(|e| {
        panic!(
            "
--------------------------------------------------------------------------
Error reading dbc test files from   {dir_display}
{e}
Make sure git submodules are up to date by running
    git submodule update --init --recursive
--------------------------------------------------------------------------
"
        )
    });

    eprintln!("Testing dbc files in directory: {dir_display}");
    for dbc_path in dir_content {
        let path = dbc_path.unwrap().path();
        if path
            .extension()
            .is_some_and(|ext| ext.eq_ignore_ascii_case("dbc"))
        {
            parse_one_file(path, snapshot, cp1251);
        }
    }
}

/// Parse a single DBC file and assert a snapshot of the result.
fn parse_one_file(path: impl AsRef<Path>, snapshot: bool, cp1251: bool) {
    let path = path.as_ref();
    eprintln!("Testing DBC file: {}", path.display());
    let file_name = path.file_stem().unwrap().to_string_lossy().to_string();
    let buffer = fs::read(path).unwrap();
    let buffer = if cp1251 {
        decode_cp1252(&buffer)
            .unwrap_or_else(|| panic!("Failed to decode {} as cp1252", path.display()))
    } else {
        Cow::Borrowed(
            std::str::from_utf8(&buffer)
                .unwrap_or_else(|_| panic!("Failed to decode {} as utf-8", path.display())),
        )
    };

    match DbcParser::parse(Rule::file, &buffer) {
        Ok(pairs) => {
            if std::env::var("SKIP_INSTA").is_err()
                && (snapshot || std::env::var("FORCE_INSTA").is_ok())
            {
                assert_debug_snapshot!(file_name, pairs);
            }
        }
        Err(e) => panic!("Failed to parse {file_name}.dbc: {e:#?}"),
    }
}
