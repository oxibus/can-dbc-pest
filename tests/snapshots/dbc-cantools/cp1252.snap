---
source: tests/snapshots.rs
---
[
    Pair {
        rule: file,
        span: Span {
            str: "VERSION \"\"\n\n\nNS_ : \n\tNS_DESC_\n\tCM_\n\tBA_DEF_\n\tBA_\n\tVAL_\n\tCAT_DEF_\n\tCAT_\n\tFILTER\n\tBA_DEF_DEF_\n\tEV_DATA_\n\tENVVAR_DATA_\n\tSGTYPE_\n\tSGTYPE_VAL_\n\tBA_DEF_SGTYPE_\n\tBA_SGTYPE_\n\tSIG_TYPE_REF_\n\tVAL_TABLE_\n\tSIG_GROUP_\n\tSIG_VALTYPE_\n\tSIGTYPE_VALTYPE_\n\tBO_TX_BU_\n\tBA_DEF_REL_\n\tBA_REL_\n\tBA_DEF_DEF_REL_\n\tBU_SG_REL_\n\tBU_EV_REL_\n\tBU_BO_REL_\n\tSG_MUL_VAL_\n\nBS_:\n\nBU_: Node1\n\n\n\n\nCM_ BU_ Node1 \" !#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~\u{7f}€‚ƒ„…†‡ˆ‰Š‹ŒŽ‘’“”•–—˜™š›œžŸ\u{a0}¡¢£¤¥¦§¨©ª«¬\u{ad}®¯°±²³´µ¶·¸¹º»¼½¾¿ÀÁÂÃÄÅÆÇÈÉÊËÌÍÎÏÐÑÒÓÔÕÖ×ØÙÚÛÜÝÞßàáâãäåæçèéêëìíîïðñòóôõö÷øùúûüýþÿ\";\nBA_DEF_  \"BusType\" STRING ;\nBA_DEF_DEF_  \"BusType\" \"\";\n\n",
            start: 0,
            end: 790,
        },
        inner: [
            Pair {
                rule: version,
                span: Span {
                    str: "VERSION \"\"\n",
                    start: 0,
                    end: 11,
                },
                inner: [
                    Pair {
                        rule: quoted_str,
                        span: Span {
                            str: "\"\"",
                            start: 8,
                            end: 10,
                        },
                        inner: [
                            Pair {
                                rule: string,
                                span: Span {
                                    str: "",
                                    start: 9,
                                    end: 9,
                                },
                                inner: [],
                            },
                        ],
                    },
                ],
            },
            Pair {
                rule: new_symbols,
                span: Span {
                    str: "NS_ : \n\tNS_DESC_\n\tCM_\n\tBA_DEF_\n\tBA_\n\tVAL_\n\tCAT_DEF_\n\tCAT_\n\tFILTER\n\tBA_DEF_DEF_\n\tEV_DATA_\n\tENVVAR_DATA_\n\tSGTYPE_\n\tSGTYPE_VAL_\n\tBA_DEF_SGTYPE_\n\tBA_SGTYPE_\n\tSIG_TYPE_REF_\n\tVAL_TABLE_\n\tSIG_GROUP_\n\tSIG_VALTYPE_\n\tSIGTYPE_VALTYPE_\n\tBO_TX_BU_\n\tBA_DEF_REL_\n\tBA_REL_\n\tBA_DEF_DEF_REL_\n\tBU_SG_REL_\n\tBU_EV_REL_\n\tBU_BO_REL_\n\tSG_MUL_VAL_\n\n",
                    start: 13,
                    end: 337,
                },
                inner: [
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "NS_DESC_",
                            start: 21,
                            end: 29,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "CM_",
                            start: 31,
                            end: 34,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "BA_DEF_",
                            start: 36,
                            end: 43,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "BA_",
                            start: 45,
                            end: 48,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "VAL_",
                            start: 50,
                            end: 54,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "CAT_DEF_",
                            start: 56,
                            end: 64,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "CAT_",
                            start: 66,
                            end: 70,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "FILTER",
                            start: 72,
                            end: 78,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "BA_DEF_DEF_",
                            start: 80,
                            end: 91,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "EV_DATA_",
                            start: 93,
                            end: 101,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "ENVVAR_DATA_",
                            start: 103,
                            end: 115,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "SGTYPE_",
                            start: 117,
                            end: 124,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "SGTYPE_VAL_",
                            start: 126,
                            end: 137,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "BA_DEF_SGTYPE_",
                            start: 139,
                            end: 153,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "BA_SGTYPE_",
                            start: 155,
                            end: 165,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "SIG_TYPE_REF_",
                            start: 167,
                            end: 180,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "VAL_TABLE_",
                            start: 182,
                            end: 192,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "SIG_GROUP_",
                            start: 194,
                            end: 204,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "SIG_VALTYPE_",
                            start: 206,
                            end: 218,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "SIGTYPE_VALTYPE_",
                            start: 220,
                            end: 236,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "BO_TX_BU_",
                            start: 238,
                            end: 247,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "BA_DEF_REL_",
                            start: 249,
                            end: 260,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "BA_REL_",
                            start: 262,
                            end: 269,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "BA_DEF_DEF_REL_",
                            start: 271,
                            end: 286,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "BU_SG_REL_",
                            start: 288,
                            end: 298,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "BU_EV_REL_",
                            start: 300,
                            end: 310,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "BU_BO_REL_",
                            start: 312,
                            end: 322,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "SG_MUL_VAL_",
                            start: 324,
                            end: 335,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: bit_timing,
                span: Span {
                    str: "BS_:\n",
                    start: 337,
                    end: 342,
                },
                inner: [],
            },
            Pair {
                rule: nodes,
                span: Span {
                    str: "BU_: Node1\n\n\n\n\n",
                    start: 343,
                    end: 358,
                },
                inner: [
                    Pair {
                        rule: node_name,
                        span: Span {
                            str: "Node1",
                            start: 348,
                            end: 353,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: comment,
                span: Span {
                    str: "CM_ BU_ Node1 \" !#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~\u{7f}€‚ƒ„…†‡ˆ‰Š‹ŒŽ‘’“”•–—˜™š›œžŸ\u{a0}¡¢£¤¥¦§¨©ª«¬\u{ad}®¯°±²³´µ¶·¸¹º»¼½¾¿ÀÁÂÃÄÅÆÇÈÉÊËÌÍÎÏÐÑÒÓÔÕÖ×ØÙÚÛÜÝÞßàáâãäåæçèéêëìíîïðñòóôõö÷øùúûüýþÿ\";",
                    start: 358,
                    end: 733,
                },
                inner: [
                    Pair {
                        rule: node_var,
                        span: Span {
                            str: "BU_ Node1",
                            start: 362,
                            end: 371,
                        },
                        inner: [
                            Pair {
                                rule: node_name,
                                span: Span {
                                    str: "Node1",
                                    start: 366,
                                    end: 371,
                                },
                                inner: [],
                            },
                        ],
                    },
                    Pair {
                        rule: quoted_str,
                        span: Span {
                            str: "\" !#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~\u{7f}€‚ƒ„…†‡ˆ‰Š‹ŒŽ‘’“”•–—˜™š›œžŸ\u{a0}¡¢£¤¥¦§¨©ª«¬\u{ad}®¯°±²³´µ¶·¸¹º»¼½¾¿ÀÁÂÃÄÅÆÇÈÉÊËÌÍÎÏÐÑÒÓÔÕÖ×ØÙÚÛÜÝÞßàáâãäåæçèéêëìíîïðñòóôõö÷øùúûüýþÿ\"",
                            start: 372,
                            end: 732,
                        },
                        inner: [
                            Pair {
                                rule: string,
                                span: Span {
                                    str: " !#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~\u{7f}€‚ƒ„…†‡ˆ‰Š‹ŒŽ‘’“”•–—˜™š›œžŸ\u{a0}¡¢£¤¥¦§¨©ª«¬\u{ad}®¯°±²³´µ¶·¸¹º»¼½¾¿ÀÁÂÃÄÅÆÇÈÉÊËÌÍÎÏÐÑÒÓÔÕÖ×ØÙÚÛÜÝÞßàáâãäåæçèéêëìíîïðñòóôõö÷øùúûüýþÿ",
                                    start: 373,
                                    end: 731,
                                },
                                inner: [],
                            },
                        ],
                    },
                ],
            },
            Pair {
                rule: attr_def,
                span: Span {
                    str: "BA_DEF_  \"BusType\" STRING ;",
                    start: 734,
                    end: 761,
                },
                inner: [
                    Pair {
                        rule: attribute_name,
                        span: Span {
                            str: "\"BusType\"",
                            start: 743,
                            end: 752,
                        },
                        inner: [
                            Pair {
                                rule: string,
                                span: Span {
                                    str: "BusType",
                                    start: 744,
                                    end: 751,
                                },
                                inner: [],
                            },
                        ],
                    },
                    Pair {
                        rule: attribute_type_string,
                        span: Span {
                            str: "STRING",
                            start: 753,
                            end: 759,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: ba_def_def,
                span: Span {
                    str: "BA_DEF_DEF_  \"BusType\" \"\";",
                    start: 762,
                    end: 788,
                },
                inner: [
                    Pair {
                        rule: attribute_name,
                        span: Span {
                            str: "\"BusType\"",
                            start: 775,
                            end: 784,
                        },
                        inner: [
                            Pair {
                                rule: string,
                                span: Span {
                                    str: "BusType",
                                    start: 776,
                                    end: 783,
                                },
                                inner: [],
                            },
                        ],
                    },
                    Pair {
                        rule: quoted_str,
                        span: Span {
                            str: "\"\"",
                            start: 785,
                            end: 787,
                        },
                        inner: [
                            Pair {
                                rule: string,
                                span: Span {
                                    str: "",
                                    start: 786,
                                    end: 786,
                                },
                                inner: [],
                            },
                        ],
                    },
                ],
            },
            Pair {
                rule: EOI,
                span: Span {
                    str: "",
                    start: 790,
                    end: 790,
                },
                inner: [],
            },
        ],
    },
]
